keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,373 @@
+an [OTLP trace provider](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/provider.go) along with a [context propagator](https://opentelemetry.io/docs/go/instrumentation/#propagators-and-context) will be configured.
+
+Also, designated http servers and clients will be wrapped with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/v0.21.0/instrumentation/net/http)
+to generate spans for sampled incoming requests and propagate context with client requests.
+
+OpenTelemetry-Go provides the [propagation package](https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/propagation.go) with which you can add custom key-value pairs known as [baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md). Baggage data will be propagated across services within contexts.
+
+### Running the OpenTelemetry Collector
+
+Although this proposal focuses on running the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector), note that any
+[OTLP-compatible](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md)
+trace collection pipeline can be used to collect OTLP data from the kubelet. As an example, the OpenTelemetry Collector can be run as a sidecar, a daemonset,
+a deployment , or a combination in which the daemonset buffers telemetry and forwards to the deployment for aggregation (e.g. tail-base sampling) and routing to
+a telemetry backend.  To support these various setups, the kubelet should be able to send traffic either to a local (on the control plane network) collector,
+or to a cluster service (on the cluster network).
+
+### Kubelet Configuration
+
+Refer to [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ccbfdd9020ba0d581aa5d880235302f8b0eb8669/specification/resource/semantic_conventions/README.md#service)
+
+```golang
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration componentbaseconfig.TracingConfiguration
+}
+
+```
+
+**NOTE** The TracingConfiguration will be defined in `kubernetes/component-base` repository
+```golang
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    // +optional
+    // URL of the collector that's running on the node.
+    // Defaults to 0.0.0.0:4317
+    URL  *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
```

Comment:
Will this sampling be "aware" of different pod attributes (like RuntimeClass, StorageClass of PVCs referred by the pod, pod security context fields that affect pod bringup, etc) and make some sort of attempt to spread samples evenly across such attributes? As a concrete example, let's say, high pod bringup latencies are observed when a certain pod annotation causes a specific IP allocation to kick in as part of CNI operations. In order to collect lots of traces specifically for pods with the special annotation, how would a cluster operator go about specifying the sampling rate?

Reply:
sampling rate will be updated in the KubeletConfiguration, and with a restart of kubelet service per node

Reply:
The sampling rate isn't aware of attributes.  For those kinds of use-cases, we generally try and solve them outside of the application itself so applications don't need complex configuration.  Right now, the best way to accomplish this would be using the tail-based sampling processor: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b90d1f0e4806f4bd5c3836e26af8f1def583f38d/processor/tailsamplingprocessor/README.md#tail-sampling-processor.  You would set a high sampling rate for the application, and then filter based on attributes you care about in the collector.